flocker/common/test/test_openstack.py

```diff
@@ -0,0 +1,142 @@
+    HttpError as KeystoneHttpError,
+)
+
+from requests import Response
+
+from .. import auto_openstack_logging
+from .._thread import NOVA_CLIENT_EXCEPTION, KEYSTONE_HTTP_ERROR
+
+
+class IDummy(Interface):
+    def return_method():
+        """
+        Return something.
+        """
+
+    def raise_method():
+        """
+        Raise something.
+        """
+
+
+@implementer(IDummy)
+class Dummy(object):
+    def __init__(self, result):
+        self._result = result
+
+    def return_method(self):
+        return self._result
+
+    def raise_method(self):
+        raise self._result
+
+
+@auto_openstack_logging(IDummy, "_dummy")
+class LoggingDummy(object):
+    def __init__(self, dummy):
+        self._dummy = dummy
+
+
+class AutoOpenStackLoggingTests(SynchronousTestCase):
+    """
+    Tests for ``auto_openstack_logging``.
+    """
+    def test_return(self):
+        """
+        Decorated methods return the value returned by the original method.
+        """
+        result = object()
+        logging_dummy = LoggingDummy(Dummy(result))
+        self.assertIs(result, logging_dummy.return_method())
+
+    def test_raise(self):
+        """
+        Decorated methods raise the same exception raised by the original
+        method.
+        """
+        result = ValueError("Things.")
+        logging_dummy = LoggingDummy(Dummy(result))
+        exception = self.assertRaises(ValueError, logging_dummy.raise_method)
+        self.assertIs(result, exception)
+
+    @capture_logging(lambda self, logger: None)
+    def test_novaclient_exception(self, logger):
+        """
+        The details of ``novaclient.exceptions.ClientException`` are logged
+        when it is raised by the decorated method and the exception is still
+        raised.
+        """
+        result = NovaClientException(
+            code=INTERNAL_SERVER_ERROR,
+            message="Some things went wrong with some other things.",
+            details={"key": "value"},
+            request_id="abcdefghijklmnopqrstuvwxyz",
+            url="/foo/bar",
+            method="POST",
+        )
+        logging_dummy = LoggingDummy(Dummy(result))
+        self.assertRaises(NovaClientException, logging_dummy.raise_method)
+
+        logged = LoggedMessage.of_type(
+            logger.messages, NOVA_CLIENT_EXCEPTION,
+        )[0]
+        assertContainsFields(
+            self, logged.message, {
+                u"code": result.code,
+                u"message": result.message,
+                u"details": result.details,
+                u"request_id": result.request_id,
+                u"url": result.url,
+                u"method": result.method,
+            },
+        )
+
+    @capture_logging(lambda self, logger: None)
+    def test_keystone_client_exception(self, logger):
+        """
+        ``keystoneclient.openstack.common.apiclient.exceptions.BadRequest`` is
+        treated similarly to ``novaclient.exceptions.ClientException``.
```

Comment:
Either refer explicitly to the test whose behaviour this test is going analogous to, or explicitly explain the expected behaviour; it isn't clear that people will be reading the test docstrings in order.